rust/agama-dbus-server/src/main.rs

```diff
@@ -0,0 +1,57 @@
+use std::{error::Error, future::pending};
+use zbus::{ConnectionBuilder, dbus_interface};
+
+struct Locale {
+    locale_id: String,
+    keyboard_id: String,
+    timezone_id: String
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Locale1")]
+impl Locale {
+    // Can be `async` as well.
+    fn list_locales(&self, locale: &str) -> Vec<(String, String)> {
+        let locales = agama_locale_data::get_languages();
+        // TODO: localization param
+        return locales.language.iter().map(|l| (l.id.clone(), locale.to_string())).collect()
+    }
+
+    fn set_locale(&mut self, locale: &str) {
+        self.locale_id = locale.to_string();
+    }
+
+    fn list_x11_keyboards(&self) -> Vec<(String, String)> {
+        let keyboards = agama_locale_data::get_keyboards();
+        return keyboards.keyboard.iter().map(|k| (k.id.clone(), k.description.clone())).collect()
+    }
+
+    fn set_x11_keyboard(&mut self, keyboard: &str) {
+        self.keyboard_id = keyboard.to_string();
+    }
+
+    fn list_timezones(&self, locale: &str) -> Vec<(String, String)> {
```

Comment:
A bad input makes a thread panic and then the service is running in an unusable state. We should send error replies instead

```console
$ ./target/debug/agama-dbus-server &
$ busctl --user call org.opensuse.Agama.Locale1 /org/opensuse/Agama/Locale1 org.opensuse.Agama.Locale1 ListTimezones s idontknow
thread 'zbus::Connection executor' panicked at 'Wrong language {language}', agama-locale-data/src/localization.rs:12:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Call failed: Connection timed out
```